README.md
oc_datadog
=====

[![CircleCI](https://circleci.com/gh/opencensus-beam/opencensus_datadog.svg?style=svg)](https://circleci.com/gh/opencensus-beam/opencensus_datadog)
[![codecov](https://codecov.io/gh/opencensus-beam/opencensus_datadog/branch/master/graph/badge.svg)](https://codecov.io/gh/opencensus-beam/opencensus_datadog)

[Opencensus][oc] integration to [DataDog][dd] traces and metrics (via dogstatsd).

## Installation

Rebar3:

```erlang
{deps, [{oc_datadog, "~> 0.1.0"}]}.
```

Mix:

```elixir
def deps do
  [
    {:oc_datadog, "~> 0.1.0"}
  ]
end
```

## Configuration

`sys.conf`:

```erlang
[
    {opencensus, [
        {reporter, {oc_reporter_datadog, []}},
        {stat, [
            {exporters, [
                {oc_stat_exporter_datadog, []}
            ]}
        }]
    ]}
].
```

Elixir:

```elixir
config :opencensus, :reporter, {:oc_reporter_datadog, []}

config :opencensus, :stat,
  exporters: [{:oc_stat_exporter_datadog, []}]
```

## License

See [LICENSE](LICENSE) file.

[oc]: https://github.com/census-instrumentation/opencensus-erlang "Opencensus Erlang"
[dd]: https://datadoghq.com "DataDog"


CHANGELOG.md
<a name="0.2.0"></a>
## 0.2.0 (2019-01-27)


#### Features

*   add dogstatsd exporter for metrics ([7b14667e](7b14667e))

#### Bug Fixes

*   automatically sanitize metric names ([4f9f9942](4f9f9942))
*   incorrect singular form in erlang:float_to_binary/2 call ([5ca1135c](5ca1135c))
*   incorrect guard on matching successful report ([b2cdcf63](b2cdcf63))


test/oc_reporter_datadog_SUITE.erl
-module(oc_reporter_datadog_SUITE).

-include_lib("opencensus/include/opencensus.hrl").
-include_lib("common_test/include/ct.hrl").
-include_lib("eunit/include/eunit.hrl").

-export([all/0, init_per_testcase/2]).
-export([test_eunit/1, test_reports_spans/1]).

all() -> [test_eunit, test_reports_spans].

init_per_testcase(_, Config) ->
    Span = #span{
              name = <<"foo">>,
              trace_id = 1,
              span_id = 1,
              start_time = wts:timestamp(),
              end_time = wts:timestamp(),
              attributes = #{foo => "bar"}
             },
    Options = oc_reporter_datadog:init([{http_client, fun mock_client/3}]),
    [{span, Span}, {options, Options} | Config].

mock_client(Address, Headers, JSON) ->
    self() ! {http, lists:flatten(Address), Headers, JSON},
    ok.

test_reports_spans(Config) ->
    Options = ?config(options, Config),
    Span = ?config(span, Config),
    oc_reporter_datadog:report([Span], Options),
    receive
        {http, "http://localhost:8126/v0.3/traces", _, JSON} ->
            Data = jsx:decode(JSON, [return_maps]),
            [[RSpan]] = Data,
            #{<<"name">> := <<"foo">>,
             <<"trace_id">> := 1,
             <<"span_id">> := 1,
             <<"type">> := <<"custom">>,
             <<"duration">> := _,
             <<"meta">> := #{<<"foo">> := <<"bar">>}} = RSpan,
            ok;
        Msg ->
            ct:fail("Unknown message: ~p", [Msg])
    after
        1000 -> ct:fail("Didn't received message in 1s")
    end.

%% EUNIT TESTS =================================================================

test_eunit(_Config) -> eunit:test(?MODULE, []).

init_return_value_contains_needed_keys_test_() ->
    Opts = oc_reporter_datadog:init([]),
    {"test that init value contains needed keys", inparallel,
     [?_assert(maps:is_key(service, Opts)),
      ?_assert(maps:is_key(host, Opts)),
      ?_assert(maps:is_key(port, Opts)),
      ?_assert(maps:is_key(type, Opts))]}.

init_sets_proper_fields_to_requested_values_test_() ->
    {"test that init sets proper fields in resulting map", inparallel,
     [init_sets(host, "foo"),
      init_sets(port, 6666),
      init_sets(service, "foo"),
      init_sets(type, "foo")]}.

init_sets(Key, Value) ->
    Opts = oc_reporter_datadog:init([{Key, Value}]),
    ?_assertEqual(Value, maps:get(Key, Opts)).


test/oc_stat_exporter_datadog_SUITE.erl
-module(oc_stat_exporter_datadog_SUITE).

-include_lib("common_test/include/ct.hrl").
-include_lib("eunit/include/eunit.hrl").

-export([all/0, init_per_testcase/2, end_per_testcase/2]).
-export([test_eunit/1, test_export/1, test_export_with_tags/1]).

all() -> [test_eunit, test_export, test_export_with_tags].

init_per_testcase(test_eunit, Config) -> Config;
init_per_testcase(_, Config) ->
    _ = application:ensure_all_started(oc_datadog),
    {ok, Socket} = gen_udp:open(8125, [{active, once}]),
    Measure = oc_stat_measure:new('datadog/test', "Test", foos),
    {ok, View} = oc_stat_view:subscribe(#{
                  name => 'datadog/test',
                  measure => Measure,
                  description => "Test",
                  tags => [foo, bar],
                  aggregation => oc_stat_aggregation_count}),
    [{udp, Socket}, {measure, Measure}, {view, View} | Config].

end_per_testcase(test_eunit, Config) -> Config;
end_per_testcase(_, Config) ->
    Socket = ?config(udp, Config),
    View = ?config(view, Config),
    oc_stat_view:unsubscribe(View),
    oc_stat_view:deregister(View),
    gen_udp:close(Socket),
    Config.

test_eunit(_Config) -> eunit:test(?MODULE, []).

test_export(Config) ->
    _ = oc_stat:record(#{}, 'datadog/test', 1),
    View = ?config(view, Config),
    Data = oc_stat_view:export(View),
    oc_stat_exporter_datadog:export([Data], []),
    receive
        {udp, _, _, _, "datadog.test:1|g"} -> ok;
        Msg ->
            ct:fail("Unknown message: ~p", [Msg])
    after
        1000 -> ct:fail("Didn't received message in 1s")
    end.

test_export_with_tags(Config) ->
    _ = oc_stat:record(#{foo => "1", bar => "2"}, 'datadog/test', 1),
    View = ?config(view, Config),
    Data = oc_stat_view:export(View),
    oc_stat_exporter_datadog:export([Data], []),
    receive
        {udp, _, _, _, "datadog.test:1|g|#bar:2,foo:1"} -> ok;
        {udp, _, _, _, "datadog.test:1|g|#foo:1,bar:2"} -> ok;
        Msg ->
            ct:fail("Unknown message: ~p", [Msg])
    after
        1000 -> ct:fail("Didn't received message in 1s")
    end.


src/oc_stat_exporter_datadog.erl
%% @author Łukasz Niemier <lukasz+opensource@niemier.pl>
%%
%% @doc Trace reporter for DataDog ([https://datadog.com]).
%%
%% == Configuration ==
%%
%% <ul>
%%  <li>`host' - address where DataDog Agent lives (default `"localhost"')</li>
%%  <li>
%%    `port' - port on which Agent listens for statsd metrics (default `8125')
%%  </li>
%% </ul>
-module(oc_stat_exporter_datadog).

-behaviour(oc_stat_exporter).

-export([export/2]).

-define(DEFAULT_HOST, "localhost").
-define(DEFAULT_PORT, 8125).

-ifdef(OTP_RELEASE).
-include_lib("kernel/include/logger.hrl").
-else.
-define(LOG_ERROR(Format, Data), error_logger:error_msg(Format, Data)).
-endif.

export(ViewData, Options) ->
    Host = proplists:get_value(host, Options, ?DEFAULT_HOST),
    Port = proplists:get_value(port, Options, ?DEFAULT_PORT),
    {ok, Socket} = gen_udp:open(0, [{active, false}]),
    Packet = build_packet(ViewData),
    ok = gen_udp:send(Socket, Host, Port, Packet),
    ok = gen_udp:close(Socket),
    ok.

build_packet(Data) when is_list(Data) ->
    List = [build_packet(Entry) || Entry <- Data],
    lists:join($\n, List);
build_packet(#{name := Name,
               ctags := CTags,
               tags := Tags,
               data := #{type := Type,
                         rows := Rows}}) ->
    Key = to_key(Name),
    List = [ build_rows(Key, Type, CTags, Tags, Row) || Row <- Rows ],
    lists:join($\n, List).

build_tags(Tags, TagsV, CTags) ->
    TagsMap = maps:merge(CTags, maps:from_list(lists:zip(Tags, TagsV))),
    TagsList = maps:to_list(TagsMap),
    Cleaned = [{Key, Value} || {Key, Value} <- TagsList, Value =/= undefined],
    build_tags(Cleaned).

build_tags([]) -> [];
build_tags(Tags) ->
    List = [[to_key(Key), $:, Value]
      || {Key, Value} <- Tags],
    ["|#", lists:join($,, List)].

to_key(Atom) when is_atom(Atom) ->
    sanitize(erlang:atom_to_list(Atom));
to_key(Value) -> sanitize(Value).

sanitize(Name) ->
    re:replace(Name, "[^[:alpha:][:digit:]_]+", ".", [global]).

build_rows(Name, Type, CTags, Tags, #{tags := TagsV, value := Value}) ->
    TagList = build_tags(Tags, TagsV, CTags),
    Metrics = [[Name, Stat] || Stat <- build_row(Type, Value, TagList) ],
    lists:join($\n, Metrics).

build_row(sum, #{count := Count,
                 mean := Mean,
                 sum := Sum}, Tags) ->
    [
     [".count:", format_num(Count), "|g", Tags],
     [".mean:", format_num(Mean), "|g", Tags],
     [".sum:", format_num(Sum), "|g", Tags]
    ];
build_row(distribution, #{buckets := Buckets} = Data, Tags) ->
    BucketRows = [ bucket_row(Bucket, Tags) || Bucket <- Buckets ],
    build_row(sum, Data, Tags) ++ BucketRows;
build_row(_Type, Value, Tags) ->
    [[$:, format_num(Value), "|g", Tags]].

bucket_row({Bound, Count}, Tags) when is_integer(Count) ->
    [
     $:, format_num(Count),
     "|g",
     tags_append(Tags, ["le:", format_num(Bound)])
    ].

tags_append([], Value) -> ["|#", Value];
tags_append(List, Value) -> [List, $,, Value].

format_num(infinity) -> <<"infinity">>;
format_num(Integer) when is_integer(Integer) ->
    erlang:integer_to_binary(Integer);
format_num(Float) when is_float(Float) ->
    erlang:float_to_binary(Float, [{decimals, 5}, compact]).


src/oc_reporter_datadog.erl
%% @author Łukasz Niemier <lukasz+opensource@niemier.pl>
%%
%% @doc Trace reporter for DataDog ([https://datadog.com]).
%%
%% == Configuration ==
%%
%% <ul>
%%  <li>`host' - address where DataDog Agent lives (default `"localhost"')</li>
%%  <li>`port' - port on which Agent listens for traces (default `8126')</li>
%%  <li>`service' - service name (default `opencensus-app')</li>
%% </ul>
-module(oc_reporter_datadog).

-behaviour(oc_reporter).

%% API exports
-export([init/1, report/2]).

-include_lib("opencensus/include/opencensus.hrl").

-ifdef(OTP_RELEASE).
-include_lib("kernel/include/logger.hrl").
-else.
-define(LOG_ERROR(Format, Data), error_logger:error_msg(Format, Data)).
-endif.

-define(TRACER_VERSION, "OC/0.1.0").

-define(DEFAULT_HOST, "localhost").
-define(DEFAULT_PORT, 8126).
-define(DEFAULT_SERVICE, <<"opencensus-app">>).
-define(DEFAULT_TYPE, <<"custom">>).

-spec init(term()) -> oc_reporter:opts().
init(Options) ->
    Host = proplists:get_value(host, Options, ?DEFAULT_HOST),
    Port = proplists:get_value(port, Options, ?DEFAULT_PORT),
    Service = proplists:get_value(service, Options, ?DEFAULT_SERVICE),
    Type = proplists:get_value(type, Options, ?DEFAULT_TYPE),
    Client = proplists:get_value(http_client, Options, fun default_client/3),
    #{host => Host,
      port => Port,
      service => Service,
      type => Type,
      client => Client}.

-spec report(nonempty_list(opencensus:span()), oc_reporter:opts()) -> ok.
report(Spans, #{
         service := Service,
         host := Host,
         port := Port,
         type := Type,
         client := Client}) ->
    Sorted = lists:sort(fun(A, B) ->
                                A#span.trace_id =< B#span.trace_id
                        end, Spans),
    Grouped = group(Sorted),
    DSpans = [[build_span(S, Service, Type) || S <- Trace] || Trace <- Grouped],

    try jsx:encode(DSpans) of
        JSON ->
            Address = io_lib:format('http://~s:~B/v0.3/traces', [Host, Port]),
            Headers = [
                       {"Datadog-Meta-Lang", "erlang"},
                       {"Datadog-Meta-Lang-Version", lang_version()},
                       {"Datadog-Meta-Lang-Interpreter", interpreter_version()},
                       {"Datadog-Meta-Tracer-Version", ?TRACER_VERSION}
                      ],
            case Client(Address, Headers, JSON) of
                ok -> ok;
                {error, {http_error, Code, Message}} ->
                    ?LOG_ERROR("DD: Unable to send spans,"
                               " DD reported an error: ~p: ~p",
                              [Code, Message]);
                {error, Reason} ->
                    ?LOG_ERROR("DD: Unable to send spans, client error: ~p",
                               [Reason])
            end
    catch
        error:Error ->
            ?LOG_ERROR("DD: Can't spans encode to json: ~p", [Error])
    end.

default_client(Address, Headers, JSON) ->
    case httpc:request(
           put,
           {Address, Headers, "application/json", JSON},
           [],
           []
          ) of
        {ok, {{_, Code, _}, _, _}} when Code >= 200, Code =< 299 ->
            ok;
        {ok, {{_, Code, _}, _, Body}} ->
            {error, {http_error, Code, Body}};
        {error, Reason} ->
            {error, Reason}
    end.

lang_version() ->
    erlang:system_info(otp_release).

interpreter_version() ->
    io_lib:format('~s-~s', [erlang:system_info(version),
                            erlang:system_info(system_architecture)]).

group([]) -> [];
group([First | Spans]) -> group(Spans, [[First]]).

group([], Acc) -> Acc;
group([#span{trace_id = Id} = Span | Spans],
      [[#span{trace_id = Id} | _] = Curr | Acc]) ->
    group(Spans, [[Span | Curr] | Acc]);
group([Span | Spans], Acc) -> group(Spans, [[Span] | Acc]).

build_span(Span, Service, Type) ->
    (optional_fields(Span))#{
      <<"trace_id">> => Span#span.trace_id rem 16#ffffffff,
      <<"span_id">> => Span#span.span_id,
      <<"name">> => Span#span.name,
      <<"resource">> => Span#span.name,
      <<"service">> => to_tag(Service),
      <<"start">> => wts:to_absolute(Span#span.start_time) * 1000,
      <<"duration">> =>
        wts:duration(Span#span.start_time, Span#span.end_time) * 1000,
      <<"type">> => to_tag(Type),
      <<"meta">> => to_meta(Span#span.attributes)}.

to_meta(Attributes) -> maps:map(fun to_tag/2, Attributes).

to_tag(Value) -> to_tag(nil, Value).

to_tag(_Name, Value) when is_function(Value) -> Value();
to_tag(_Name, Value) when is_list(Value) -> list_to_binary(Value);
to_tag(_Name, Value) -> Value.

optional_fields(Span) ->
    lists:foldl(fun(Field, Acc) ->
                       case span_field(Field, Span) of
                           undefined -> Acc;
                           Value -> maps:put(Field, Value, Acc)
                       end
               end, #{}, [<<"error">>, <<"parent_id">>]).

span_field(<<"error">>, #span{status = undefined}) -> undefined;
span_field(<<"error">>, #span{status = #status{code = 0}}) -> undefined;
span_field(<<"error">>, _) -> 1;

span_field(<<"parent_id">>, #span{parent_span_id = ParentId}) -> ParentId.
